Added ksql and jdbc realtime data streaming in grafana

Hsankesara opened the pull request:
<!--
 Before you open the request please review the following guidelines and tips to help it be more easily integrated:

 - Describe the scope of your change - i.e. what the change does.
 - Describe any known limitations with your change.
 - Please run any tests or examples that can exercise your modified code.

 Thank you for contributing! We will try to test and integrate the change as soon as we can, but be aware we have many GitHub repositories to manage and can't immediately respond to every request. There is no need to bump or check in on a pull request (it will clutter the discussion of the request).

 Also don't be worried if the request is closed or not integrated sometimes the priorities of maintainers might not match the priorities of the pull request. Don't fret, the open source community thrives on forks and GitHub makes it easy to keep your changes in a forked repo.
 -->

**Description of the change**

Added KSQL server and another JDBC connector to stream high-frequency data in Grafana. 

<!-- Describe the scope of your change - i.e. what the change does. -->

**Benefits**

It would help us visualize high frequency like smartphone activity and heart rate in the dashboard.

blootsvoets commented on helmfile.d/40-realtime-dashboard.yaml:
```diff
@@ -0,0 +1,38 @@
+  kubeContext: {{ .Values.kubeContext }}
+
+releases:
+  - name: radar-jdbc-connector-agg
+    chart: radar/radar-jdbc-connector
+    version: {{ .Values.radar_jdbc_connector_agg._chart_version }}
+    installed: {{ .Values.radar_jdbc_connector_agg._install }}
+    values:
+      - {{ .Values.radar_jdbc_connector_agg | toYaml | indent 8 | trim }}
+    set:
+      - name: kafka_num_brokers
+        value: {{ .Values.kafka_num_brokers }}
+      - name: timescaledb.password
+        value: {{ .Values.timescaledb_password }}
+      - name: timescaledb.username
+        value: postgres
+      - name: timescaledb.database_name
+        value: {{ .Values.timescaledb_db_name }}
+
```
There is already a `radar-jdbc-connector`, a second one is not needed.

Hsankesara replied:
Hi Joris, the reason behind having another JDBC connector (as @yatharthranjan mentioned to me) is that because of different cases for the keys, etc in KSQL. by default KSQL converts everything to uppercase (even if the source topic has lower case), So unless we want to quote ("") everything in all the queries (to allow for case support), we need a separate connector. It is not required, it would be convenient to have.

blootsvoets replied:
In that case, please add the config needed to enable this behaviour. Please note that the jdbc connector has been updated and the values are now called differently.

blootsvoets commented on etc/base.yaml:
```diff
@@ -442,6 +442,37 @@ radar_push_endpoint:
     consumerSecret: ""
 
+# --------------------------------------------------------- 40-realtime-analyses.yaml ---------------------------------------------------------
+
+radar_jdbc_connector_agg:
+  _install: false
+  _chart_version: 0.1.6
+  replicaCount: 1
+  timescaledb:
+    host: timescaledb-postgresql-hl
+
```
This is not needed

blootsvoets commented on etc/base.yaml:
```diff
@@ -442,6 +442,37 @@ radar_push_endpoint:
+ksql_server:
+  _install: false
+  replicaCount: 1
+  servicePort: 8088
+  kafka:
+    bootstrapServers: PLAINTEXT://cp-kafka:9092
+  cp-schema-registry: 
+    url: http://cp-schema-registry:8081
+  ksql:
+    headless: false
+  configurationOverrides:
+    "ksql.server.url": "http://0.0.0.0:8088"
+    "ksql.advertised.listener": "http://ksql-server:8088"
+    "ksql.logging.processing.topic.auto.create": true
+    "ksql.logging.processing.stream.auto.create": true
+    "ksql.service.id": radar_default_
+    "ksql.persistence.default.format.key": AVRO
+    "ksql.persistence.default.format.value": AVRO
+    "ksql.sink.replicas": 3
+    "ksql.streams.replication.factor": 3
+    "ksql.internal.topic.replicas": 3
+
```
Preferably move some sensible defaults to the `etc/cp-ksql-server/values.yaml` file.

blootsvoets reviewed (commented):
Adding KSQL looks fine to me. The radar-jdbc-connector does not need to be duplicated though.

Please see for reference also
https://github.com/IMI-H2O/h2o-kubernetes/blob/main/etc/outcomes-ksql-transform/values.yaml
https://github.com/IMI-H2O/h2o-kubernetes/blob/main/etc/production.yaml.gotmpl

Hsankesara commented:
Hi @blootsvoets,

I have added sink properties for `radar-jdbc-connector-agg`.  I created `../etc/radar-jdbc-connector-agg/values.yaml` and added sink properties there. I mimicked the config we have in realtime-dashboard in the Radar-Docker. 
One config I cannot add is `transforms.timestamp.fields=TIME`, because radar jdbc helm chart has no option of defining it ([https://github.com/RADAR-base/radar-helm-charts/blob/b93bae0d0fb124c44f0bab2c0b345[…]d6693ba86d/charts/radar-jdbc-connector/templates/configmap.yaml](https://github.com/RADAR-base/radar-helm-charts/blob/b93bae0d0fb124c44f0bab2c0b3459d6693ba86d/charts/radar-jdbc-connector/templates/configmap.yaml#L48))

blootsvoets commented:
I'm wondering now, would it be an option to just quote the variables in ksqlDB? we use ` `` ` in our stack. Otherwise, could you make a PR in radar-helm-charts to make the timestamp variables configurable? When referencing multiple tables, you only need to quote the field name, e.g.
```sql
SELECT o.`value`
FROM observation o
EMIT CHANGES;
```

yatharthranjan commented:
@blootsvoets Agreed if resources are constrained, then makes sense to have a single one but on the flip side, the current connector deployments are standalone (and not distributed), hence they do not scale, so in that sense having 2 would mean a better level of scalability if the resources are available. How about we keep this disabled by default and usual instruction would suggest to use the main jdbc-connector with quoting in KSQL queries but the helmfile does add a new connector (disabled by default) and can be enabled and configured in the production.yaml for more power users?

But if that is too complicated, we can leave it out altogether from the main repo.

blootsvoets commented:
If you make a PR for radar-helm-charts that’s fine as well, if you need to configure the mergeKey property.

Hsankesara commented:
Sounds good, will make a PR there.

Hsankesara commented:
Sounds good, will make a PR there.
